extra/lrshift~/lrshift~.c

```diff
@@ -7,68 +7,67 @@ static t_class *lrshift_tilde_class;
 typedef struct _lrshift_tilde
 {
     t_object x_obj;
-    int x_n;
-    t_float x_f;
+    t_float  x_n;
+    t_float  x_f;
 } t_lrshift_tilde;
 
-static t_int *leftshift_perform(t_int *w)
+static t_int *lrshift_perform(t_int *w)
 {
-    t_sample *in = (t_sample *)(w[1]);
-    t_sample *out= (t_sample *)(w[2]);
-    int n = (int)(w[3]);
-    int shift = (int)(w[4]);
-    in += shift;
-    n -= shift;
-    while (n--)
-        *out++ = *in++;
-    while (shift--)
-        *out++ = 0;
-    return (w+5);
-}
+    t_lrshift_tilde *x = (t_lrshift_tilde *) w[1];
+    t_sample *in = (t_sample *)(w[2]);
+    t_sample *out= (t_sample *)(w[3]);
+    int n = (int)(w[4]);
+    int shift = (int)x->x_n;
+    t_sample *ptr,
+        *out_end = out + n - 1,
+        *in_end = in + n - 1;
```

Comment:
`out_end` is shadowed in the else-block below. `in_end` is only really necessary for the if-block.

In this case, there is no need to calculate any pointers outside the if-else-blocks at all:
```
    if (shift >= 0)  /* shift left, shift is a postive int or zero */
    {
        if (shift > n)
            shift = n;
        n -= shift;
        in += shift;
        while (n--)
            *out++ = *in++;
        while (shift--)
            *out++ = 0.0f;
    }
    else /* shift right, shift is a negative int */
    {
        if (shift < -n)
            shift = -n;
        out += n - 1;
        in += n + shift - 1;
        n += shift;
        while (n--)
            *out-- = *in--;
        while (shift++)
            *out-- = 0.0f;
    }
```
Note: I didn't test this. Please make sure I haven't made a typo :-)

---

As a rule of thumb: first write the most natural soluation for each branch and only then see if there actually is common code that can be hoisted outside the loop.

Also, by convention "end" pointers point to one past the last element, so `out_end` should rather be `out + n`. Consequently, you would use `ptr < out_end